### PR TITLE
pageserver: separate metadata and data pages in DatadirModification

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Currently, DatadirModification keeps a key-indexed map of all pending writes, even though we (almost) never need to read back dirty pages for anything other than metadata pages (e.g. relation sizes).

Related: https://github.com/neondatabase/neon/issues/6345

## Summary of changes

- commit() modifications before ingesting database creation wal records, so that they are guaranteed to be able to get() everything they need directly from the underlying Timeline.
- Split dirty pages in DatadirModification into pending_metadata_pages and pending_data_pages.  The data ones don't need to be in a key-addressable format, so they just go in a Vec instead.
- Special case handling of zero-page writes in DatadirModification, putting them in a map which is flushed on the end of a WAL record.  This handles the case where during ingest, we might first write a zero page, and then ingest a postgres write to that page.  We used to do this via the key-indexed map of writes, but in this PR we change the data page write path to not bother indexing these by key.

My least favorite thing about this PR is that I needed to change the DatadirModification interface to add the on_record_end call.  This is not very invasive because there's really only one place we use it, but it changes the object's behaviour from being clearly an aggregation of many records to having some per-record state.  I could avoid this by implicitly doing the work when someone calls set_lsn or commit -- I'm open to opinions on whether that's cleaner or dirtier.

## Performance

There may be some efficiency improvement here, but the primary motivation is to enable an earlier stage of ingest to operate without access to a Timeline.  The `pending_data_pages` part is the "fast path" bulk write data that can in principle be generated without a Timeline, in parallel with other ingest batches, and ultimately on the safekeeper.

`test_bulk_insert` on AX102 shows approximately the same results as in the previous PR #8591:

```
------------------------------ Benchmark results -------------------------------
test_bulk_insert[neon-release-pg16].insert: 23.577 s
test_bulk_insert[neon-release-pg16].pageserver_writes: 5,428 MB
test_bulk_insert[neon-release-pg16].peak_mem: 637 MB
test_bulk_insert[neon-release-pg16].size: 0 MB
test_bulk_insert[neon-release-pg16].data_uploaded: 1,922 MB
test_bulk_insert[neon-release-pg16].num_files_uploaded: 8 
test_bulk_insert[neon-release-pg16].wal_written: 1,382 MB
test_bulk_insert[neon-release-pg16].wal_recovery: 18.264 s
test_bulk_insert[neon-release-pg16].compaction: 0.052 s
```

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/pgdatadir_mapping.rs [hunk 1]
```diff
@@ -1917,7 +1913,27 @@ impl<'a> DatadirModification<'a> {
     }
 
     fn put(&mut self, key: Key, val: Value) {
-        let values = self.pending_updates.entry(key).or_default();
+        if key.is_rel_block_key() || key.is_slru_block_key() {
+            self.put_data(key, val)
+        } else {
+            self.put_metadata(key, val)
+        }
```
Comment (VladLazar) [inline pageserver/src/pgdatadir_mapping.rs]: The distinction between data and metadata is fuzzy for me. Could you lift the "is key data predicate" into a function and add a doc comment which explains the distinction?

Comment (VladLazar) [reply pageserver/src/pgdatadir_mapping.rs]: My confusion comes from the SLRU keys. They are technically metadata, but we treat them as data.

File: pageserver/src/pgdatadir_mapping.rs [hunk 2]
```diff
@@ -1914,25 +1952,26 @@ impl<'a> DatadirModification<'a> {
...
+            self.put_data(key.to_compact(), val)
         } else {
-            self.put_metadata(key, val)
+            self.put_metadata(key.to_compact(), val)
         }
     }
 
-    fn put_data(&mut self, key: Key, val: Value) {
-        // TODO: old code had logic for eliding multiple updates at the same lsn to the same key: was that really
-        // needed for data pages?
```
Comment (VladLazar) [inline pageserver/src/pgdatadir_mapping.rs]: So the elisions were for zero pages?

File: pageserver/src/pgdatadir_mapping.rs [hunk 3]
```diff
@@ -1183,6 +1195,29 @@ impl<'a> DatadirModification<'a> {
         Ok(())
     }
 
+    pub(crate) fn put_rel_page_image_zero(&mut self, rel: RelTag, blknum: BlockNumber) {
+        self.pending_zero_data_pages
+            .insert(rel_block_to_key(rel, blknum).to_compact());
```
Comment (VladLazar) [inline pageserver/src/pgdatadir_mapping.rs:1200]: Why not count these zero pages towards `pending_bytes`?

**Review by VladLazar — COMMENTED** [review]
Looks good. Thanks for splitting this up nicely.
I'm a bit confused about data vs metadata pages though.

File: pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs [hunk 4]
```diff
@@ -327,6 +327,20 @@ pub(super) async fn handle_walreceiver_connection(
                         let mut decoded = DecodedWALRecord::default();
                         decode_wal_record(recdata, &mut decoded, modification.tline.pg_version)?;
 
+                        if decoded.is_dbase_create_copy(timeline.pg_version)
+                            && uncommitted_records > 0
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:347]: You're right... I think the strictest check would be if `!modification.pending_data_pages.is_empty()`, but I'd prefer to keep this condition similar to the other commit() checks in this function which just check the uncommitted records count.

File: pageserver/src/pgdatadir_mapping.rs (hunk 3 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs:1216]: Hmm, this was an oversight: we can account for this properly if we add here and then subtract if we later remove the zero page: 562b3706e

(While writing this I also have quite a strong urge to reprezent zero pages in a more compact way than literally writing 8192 bytes of zeros...)

File: pageserver/src/pgdatadir_mapping.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs]: Right.  I checked with the compute team that they don't expect postgres itself to ever write the same (key, lsn) twice: these are just originating from the synthetic zero writes the pageserver generates.

File: pageserver/src/pgdatadir_mapping.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs]: Good call: clarified the data/metadata thing in a helper function here: 68a6c46d1

File: pageserver/src/pgdatadir_mapping.rs [hunk 5]
```diff
@@ -1066,13 +1082,28 @@ impl<'a> DatadirModification<'a> {
...
+    /// In this context, 'metadata' means keys that are only read by the pageserver internally, and 'data' means
+    /// keys that represent literal blocks that postgres can read.  So data includes relation blocks and
+    /// SLRU blocks, which are read directly by postgres, and everything else is considered metadata.
+    ///
+    /// The distinction is important because data keys are handled on a fast path where dirty writes are
+    /// not readable until this modification is committed, whereas metadata keys are visible for read
+    /// via [`Self::get`] as soon as their record has been ingested.
+    fn is_data_key(key: &Key) -> bool {
+        key.is_rel_block_key() || key.is_slru_block_key()
+    }
```
Comment (problame) [inline pageserver/src/pgdatadir_mapping.rs:1105]: What about auxv2?

File: pageserver/src/pgdatadir_mapping.rs [hunk 6]
```diff
@@ -1917,7 +1969,30 @@ impl<'a> DatadirModification<'a> {
...
+        self.pending_bytes += val_serialized_size;
+        self.pending_data_pages
+            .push((key, self.lsn, val_serialized_size, val))
+    }
+
+    fn put_metadata(&mut self, key: CompactKey, val: Value) {
+        let values = self.pending_metadata_pages.entry(key).or_default();
         // Replace the previous value if it exists at the same lsn
         if let Some((last_lsn, last_value_ser_size, last_value)) = values.last_mut() {
             if *last_lsn == self.lsn {
```
Comment (problame) [inline pageserver/src/pgdatadir_mapping.rs:1998]: Likely pre-existing bug, but, don't we need to adjust `self.pending_bytes` if we replace the value?

File: pageserver/src/pgdatadir_mapping.rs [hunk 7]
```diff
@@ -1879,33 +1924,40 @@ impl<'a> DatadirModification<'a> {
...
+                            "unexpected pending WAL record"
+                        )))
+                    };
+                }
             }
         }
+
+        // Metadata page cache miss, or we're reading a data page.
         let lsn = Lsn::max(self.tline.get_last_record_lsn(), self.lsn);
         self.tline.get(key, lsn, ctx).await
```
Comment (problame) [inline pageserver/src/pgdatadir_mapping.rs:1962]: Maybe I'm mis-remembering the invariants we discussed in the call about these refactorings.

But, can we assert here that `self.pending_data_pages` doesn't have an entry for `key`?

If not, why are we allowed to bypass `self.pending_data_pages`? Either way, warrants a mention in this method's doc comment. 

Also please consider `self.pending_zero_data_pages` in mentions / assertions that are added, same concern applies to them.

**Review by problame — COMMENTED** [review]
I'm not sure the `zero_data_pages` is worth it. You can just use a `Bytes::from_static(ZERO_PAGE)`.

File: pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs [hunk 8]
```diff
@@ -324,9 +339,28 @@ pub(super) async fn handle_walreceiver_connection(
...
+                            )
+                            .await?;
+                        }
+
                         // Ingest the records without immediately committing them.
                         let ingested = walingest
-                            .ingest_record(recdata, lsn, &mut modification, &mut decoded, &ctx)
+                            .ingest_record(decoded, lsn, &mut modification, &ctx)
                             .await
                             .with_context(|| format!("could not ingest record at {lsn}"))?;
```
Comment (problame) [inline pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:365]: Would like to see an assertion inside `walingest` or `DatadirModification` asserting that `DatadirModification` must be empty when ingesting a record of type `is_dbase_create_copy()`.

File: pageserver/src/pgdatadir_mapping.rs (hunk 5 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs:1105]: Those are considered metadata in this scheme: data are only literal postgres pages, rather than pages that store something that one needs a pageserver to understand.

File: pageserver/src/pgdatadir_mapping.rs (hunk 7 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs:1962]: > can we assert here that self.pending_data_pages doesn't have an entry for key?

We can indeed: 4e9cd3c58 -- initially I had skipped this because it's an O(N) thing, but on reflection we only hit this path during legacy db creates, so a debug-mode assertion is fine here.  Also asserting on pending_zero_data_pages at the same time.

File: pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs (hunk 8 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/walreceiver/walreceiver_connection.rs:365]: a189f989b -- this is a bit awkward because we're immediately asserting something that is done one line before calling the function, but I agree that it makes us safer against future changes.

File: pageserver/src/pgdatadir_mapping.rs (hunk 6 above)
Comment (jcsp) [reply pageserver/src/pgdatadir_mapping.rs:2020]: Good spot.  Fixed in 70c9e2320

**Comment by jcsp** [comment]
When merging `main`, I noticed that changes from #8648 were calling `is_valid_key_on_write_path` during ingest at some points where we can't do that any more (because they get converted down to i128 earlier, in non-fallible functions).

Rather than making lots more functions fallible, I'm inclined to leave those calls out (that's what I've done in the merge commit), and rely on the single check in Timeline::put, which all writes will hit eventually.  @skyzh does that sound okay to you?

**Comment by skyzh** [comment]
sounds good :)

**Review by VladLazar — APPROVED** [review]
lgtm

**Comment by skyzh** [comment]
> and rely on the single check in Timeline::put, which all writes will hit eventually

it seems that Timeline::put has `#[cfg(test)]`, it's only used in test cases? 🤣

**Comment by problame** [comment]
Re-reviewed this change, just to be completely sure we only call `DatadirModification::get` with metadata keys except the one legacy edge case. Can confirm it's true.

I'm wondering how much effort / code churn it would be to require a different `Key` type argument for `DatadirModification::get` that, at compile time, ensures the key is metadata.

Like, a wrapper type

```rust
struct MetadataKey(Key);

impl From<MetadataKey> for Key { ... }
```

and change all the constants like `DBDIR_KEY` and functions like `rel_size_to_key` to return `MetadataKey`.
